src/segment/event.rs

```diff
@@ -39,6 +39,7 @@ pub(crate) struct WriterInfo {
 
 #[derive(Debug)]
 pub(crate) struct PendingEvent {
+    pub(crate) scoped_segment: Option<ScopedSegment>,
     pub(crate) routing_key: Option<String>,
```

Comment:
Rather than have two options which are mutually exclusive, you could have an enum which has either a segment or a routing key.